src/librustc/traits/error_reporting.rs

```diff
@@ -1762,9 +1788,11 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         // - `BuiltinDerivedObligation` with `impl std::future::Future` (A)
         // - `BindingObligation` with `impl_send (Send requirement)
         //
-        // The first obligations in the chain can be used to get the details of the type that is
-        // captured but the entire chain must be inspected to detect this case.
+        // The first obligation in the chain is the most useful and has the generator that captured
+        // the type. The last generator has information about where the bound was introduced. At
+        // least one generator should be present for this diagnostic to be modified.
         let mut generator = None;
```

Comment:
Maybe refactor this logic into a method? (the current one is growing fairly large)

Reply:
Fixed!